beancount-parser/src/lib.rs

```diff
@@ -82,20 +82,50 @@ lazy_static! {
 pub struct BeancountParser;
 
 #[derive(Debug)]
-struct ParseState {
+struct ParseState<'i> {
     root_names: HashMap<bc::AccountType, String>,
+
+    // Tags added via "pushtag" instruction. Use HashMap instead of HashSet because
+    // spec allows mutliple pushing of the same tag (ie requires an equal number of pops).
+    pushed_tags: HashMap<&'i str, u16>,
 }
 
-impl ParseState {
+impl<'i> ParseState<'i> {
     fn new() -> Self {
         use bc::AccountType::*;
         ParseState {
             root_names: [Assets, Liabilities, Equity, Income, Expenses]
                 .iter()
                 .map(|ty| (*ty, ty.default_name().to_string()))
                 .collect(),
+            pushed_tags: HashMap::new(),
+        }
+    }
+
+    fn push_tag(&mut self, tag: &'i str) {
+        if let Some(count) = self.pushed_tags.get_mut(tag) {
+            *count += 1;
+        } else {
+            self.pushed_tags.insert(tag, 1);
         }
     }
+
+    fn pop_tag(&mut self, tag: &str) -> Result<(), String> {
+        if let Some(count) = self.pushed_tags.get_mut(tag) {
+            if *count == 1 {
+                self.pushed_tags.remove(tag);
+            } else {
+                *count -= 1;
+            }
+            Ok(())
+        } else {
+            Err(format!("Attempting to pop absent tag: '{}'", tag))
+        }
```

Comment:
I would personally prefer to use a `match self.pushed_tags.get_mut(tag) { ... }` here, but `if else` is OK too.

And maybe `if *count <= 1` to ensure we don't decrement a zero value (which probably shouldn't be possible, but just to be safe)?